File: tools/improve/tests/devstral-small-2/maudec/cpp/stack.cpp
//
//	<comment from="devstral-small-2">
//	I'll simplify and improve the efficiency of the code while preserving semantics. Here are the key improvements:
//	</comment>
//

#include <compare>
#include <memory>
#include <tuple>
#include <variant>

// Simplified smart_ptr - removed redundant constructors since std::shared_ptr's are sufficient
template <typename T>
class smart_ptr : public std::shared_ptr<T>
{
public:
    using std::shared_ptr<T>::shared_ptr;
    
    auto operator<=>(const smart_ptr<T>& other) const {
        T* left = this->get(), *right = other.get();
        
        if (left == right)
            return std::strong_ordering::equivalent;
        if (left && right)
            return *left <=> *right;
        
        return std::operator<=>(*this, other);
    }
    
    bool operator==(const smart_ptr<T>& other) const {
        return (*this <=> other) == 0;
    }
    
    bool operator<=(const smart_ptr<T>& other) const {
        return (*this <=> other) <= 0;
    }
};

// Simplified Stack variant - removed redundant monostate in variant
struct Stack : public std::variant<
    std::monostate,
    std::tuple<int, smart_ptr<Stack>>
> {
    using std::variant<std::monostate, std::tuple<int, smart_ptr<Stack>>>::variant;
    
    enum Variant {
        MT,
        PUSH,
        ERR,
    };
};

// Improved pop function - more efficient with direct access
Stack pop(const Stack& a1) {
    if (a1.index() == Stack::MT) {
        return Stack(std::in_place_index<Stack::MT>);
    }
    return *std::get<1>(std::get<Stack::PUSH>(a1));
}

// Improved top function - more efficient with direct access
int top(const Stack& a1) {
    return std::get<0>(std::get<Stack::PUSH>(a1));
}

// Improved empty function - more concise
bool empty(const Stack& S) {
    return S.index() == Stack::MT;
}

File: tools/improve/tests/devstral-small-2/maudec/rust/stack.rs
//
//	<comment from="devstral-small-2">
//	I'll simplify and improve the efficiency of the Rust code while preserving the semantics and keeping the original function signatures.
//	</comment>
//

#[derive(Clone, PartialEq, PartialOrd, Debug)]
enum Stack<i32> {
    Mt,
    Push(i32, Box<Stack<i32>>),
    Err,
}

fn pop(a1: &Stack<i32>) -> Stack<i32> {
    match a1 {
        Stack::Mt => Stack::Mt,
        Stack::Push(_, s) => *s.clone(),
        _ => panic!("case not covered"),
    }
}

fn top(a1: &Stack<i32>) -> i32 {
    match a1 {
        Stack::Push(e, _) => *e,
        _ => panic!("case not covered"),
    }
}

fn empty?(s: &Stack<i32>) -> bool {
    matches!(s, Stack::Mt)
}